Image controller

theGreyPilgrim42 opened the pull request:
Added a second changeScene method to the SceneController because the album view doesn't have an event to pass when an image is to be edited. 
Added an editImage method to change the Controller if exactly one image is selected to be edited.
The edit view is still basic but should work once we have an ExifService. I took it upon myself to display the content vertically (image above the metadata) because with a horizontal alignment the image will be relativly small. This can still be reverted if it looks too bad.
Please check it out and see if it works for you too.

theGreyPilgrim42 committed:
Add image view to edit an image's metadata

theGreyPilgrim42 committed:
Add ImageController

theGreyPilgrim42 committed:
Add changeScene method

Because the AlbumController will not use an Event to change scenes the
current changeScene method can't be used. A new changeScene method with
a Scene instead of an Event as a parameter solves this.

theGreyPilgrim42 committed:
Add empty method and set ImageController as view controller

theGreyPilgrim42 committed:
Add edit button to album controller

theGreyPilgrim42 committed:
Add editImage method

theGreyPilgrim42 committed:
Fix column constraint usage

theGreyPilgrim42 committed:
Update initializeImage method to display metadata

theGreyPilgrim42 committed:
Change layout of image view

The horizontal layout with the image and the metadata description next
to each other only allowed a small image.

A vertical layout allows the image to be bigger while still having
enough space to display metadata descriptions.

theGreyPilgrim42 committed:
Refactor warnUser to alertUser method

By passing the AlertType as an argument the alertUser method can be
used for all kinds of alerts.

theGreyPilgrim42 committed:
Fix minor code smell

theGreyPilgrim42 committed:
Fix instanceof check and cast with pattern matching

joshua-lehmann commented on src/main/java/ch/hftm/controllers/AlbumController.java:
```diff
@@ -275,6 +272,41 @@ public void initializeAlbum() {
+    public void editImage() {
+        String title = "";
+        String content = "";
+        if (imageSelection.size() == 1) {
+            Image imageToEdit = imageSelection.get(0);
+            Scene scene = albumPane.getScene();
+
+            SceneController sceneController = new SceneController();
+            try {
+                sceneController.changeScene(scene, "image", imageToEdit);
+            } catch (IOException e) {
+                log.error(String.format("The image %s could not be edited: %s", imageToEdit.getFileName(), e.getMessage()));
+            } finally {
+                imageSelection.clear();
+            }
+        } else if (imageSelection.isEmpty()) {
+            content = "You need to select an image you want to edit.";
+            alertUser(title, content, AlertType.WARNING);
+        } else {
+            content = "Only one image at a time can be edited.";
+            alertUser(title, content, AlertType.WARNING);
+        }
```
Again I would propose similar refactoring, if you are interested to know the general principle of simplifying those kind of if/else if/else statements you can read this article [Guard Clause](https://deviq.com/design-patterns/guard-clause)
```suggestion
        if (imageSelection.isEmpty()) {
            content = "You need to select an image you want to edit.";
            alertUser(title, content, AlertType.WARNING);
            return;
        }
        if (imageSelection.size() > 1) {
            content = "Only one image at a time can be edited.";
            alertUser(title, content, AlertType.WARNING);
            return;
        }

        Image imageToEdit = imageSelection.get(0);
        Scene scene = albumPane.getScene();
        SceneController sceneController = new SceneController();
        try {
            sceneController.changeScene(scene, "image", imageToEdit);
        } catch (IOException e) {
            log.error(String.format("The image %s could not be edited: %s", imageToEdit.getFileName(), e.getMessage()));
        } finally {
            imageSelection.clear();
        }
```

joshua-lehmann commented on src/main/java/ch/hftm/controllers/SceneController.java:
```diff
@@ -31,4 +31,20 @@ public <T> void changeScene(Event event, String sceneName, T data) throws IOExce
+        stage = (Stage) scene.getWindow();
+        stage.setScene(new Scene(root));
+        stage.show();
+    }
```
Actually this makes more sense to pass the scene instead of the event, I refactored the method which opens the album view from the library. Now only one changeScene method is used. I pushed the changes.

theGreyPilgrim42 replied:
Adding the preview features results in errors on my setup. Apparently preview features can only be enabled at source level 19. Can't we use an if-else statement instead?

joshua-lehmann committed:
#12 Improved the changeScene method to only one method used

joshua-lehmann reviewed (changes requested):
Please validate the changes I made and already pushed. And then refactor the one method and then you can merge it.

joshua-lehmann committed:
#12 Allow preview features to allow switch statement in SceneController.java

theGreyPilgrim42 committed:
Revert "#12 Allow preview features to allow switch statement in SceneController.java"

This reverts commit 180850e1c6b76f00b0217f0a81900ed9d800e86a.

theGreyPilgrim42 committed:
Refactor switch to if-else statement

The switch statement would only work with the preview feature but
VSCode doesn't support it correctly.

theGreyPilgrim42 committed:
Fix addImage method

After adding an image a case occured where a selected image couldn't be
edited, because more than one image was selected. Clearing the image
selection after adding an image resolves this issue.

joshua-lehmann reviewed (approved):
Everything looks good now
